pkg/controller/controller.go

```diff
@@ -1081,47 +1126,114 @@ func (p *csiProvisioner) getPVCSource(ctx context.Context, claim *v1.PersistentV
 	return volumeContentSource, nil
 }
 
+func (p *csiProvisioner) IsGranted(ctx context.Context, namespace, name string, claim *v1.PersistentVolumeClaim) (bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CrossNamespaceSourceProvisioning) {
+		return false, nil
+	}
+
+	// Get all ReferenceGrants in snapshot's namespace
+	referenceGrants, err := p.referenceGrantLister.ReferenceGrants(namespace).List(labels.Everything())
+	if err != nil {
+		return false, fmt.Errorf("error getting ReferenceGrants in %s namespace from api server: %v", namespace, err)
+	}
+
+	var allowed bool
+	// Check that accessing to {namespace}/{name} is allowed.
+	for _, grant := range referenceGrants  {
+		var validFrom bool
+		for _, from := range grant.Spec.From {
+			if from.Group == "" && from.Kind == pvcKind && string(from.Namespace) == claim.Namespace {
+				validFrom = true
+				break
+			}
+		}
+		// Skip unrelated policy by checking From field
+		if !validFrom {
+			continue
+		}
+
+		for _, to := range grant.Spec.To {
+			if to.Group != snapshotAPIGroup || to.Kind != snapshotKind {
+				continue
+			}
+			if to.Name == nil || string(*to.Name) == "" || string(*to.Name) == name {
+				allowed = true
+				break
+			}
+		}
+
+		if allowed {
+			break
+		}
+	}
+
+	if !allowed {
+		return false, nil
+	}
+
+	return true, nil
+}
+
 // getSnapshotSource verifies DataSource.Kind of type VolumeSnapshot, making sure that the requested Snapshot is available/ready
 // returns the VolumeContentSource for the requested snapshot
 func (p *csiProvisioner) getSnapshotSource(ctx context.Context, claim *v1.PersistentVolumeClaim, sc *storagev1.StorageClass) (*csi.VolumeContentSource, error) {
-	snapshotObj, err := p.snapshotClient.SnapshotV1().VolumeSnapshots(claim.Namespace).Get(ctx, claim.Spec.DataSource.Name, metav1.GetOptions{})
+	return p.getSnapshotSourceInternal(ctx, claim, sc, claim.Namespace, claim.Spec.DataSource.Name)
+}
+
+// getSnapshotSourceDataSourceRef2 verifies DataSourceRef2.Kind of type VolumeSnapshot, making sure that the requested Snapshot is available/ready
+// returns the VolumeContentSource for the requested snapshot
+func (p *csiProvisioner) getSnapshotSourceDataSourceRef2(ctx context.Context, claim *v1.PersistentVolumeClaim, sc *storagev1.StorageClass) (*csi.VolumeContentSource, error) {
+	if claim.Spec.DataSourceRef2.Namespace == "" {
+		return p.getSnapshotSourceInternal(ctx, claim, sc, claim.Namespace, claim.Spec.DataSourceRef2.Name)
+	}
+
+	if ok, err := p.IsGranted(ctx, claim.Spec.DataSourceRef2.Namespace, claim.Spec.DataSourceRef2.Name, claim); err != nil || !ok {
```

Comment:
Do we check somewhere if the PVC namespace is the same as the VolumeSnapshot namespace?  If they are the same, then referenceGrant is not needed.

Reply:
I'm not sure I get the context here. Agree that ReferenceGrant is only useful/necessary to enable cross-namespace references. Same namespace references should not require a ReferenceGrant.